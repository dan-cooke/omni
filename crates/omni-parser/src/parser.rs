lalrpop_mod!(pub omni);

use lalrpop_util::lalrpop_mod;

#[cfg(test)]
mod tests {

    use crate::lexer::Lexer;
    use omni_ast::{Def, File, Ident, Prop, ResourceOperation, Value};

    use super::*;

    #[test]
    pub fn _00_simple_parsing_works() {
        let source = std::fs::read_to_string("test-schemas/00_prelude/00_prelude.omni").unwrap();
        let lexer = Lexer::new(&source);
        let parser = omni::OmniFileParser::new();

        let expected_ast = File {
            imports: None,

            body: vec![
                Def::Service {
                    id: Ident {
                        id: String::from("JokeService"),
                    },
                    body: vec![Prop {
                        key: String::from("version"),
                        value: Value::String(String::from("1.0.0")),
                        decorators: None,
                    }],
                },
                Def::Read(ResourceOperation {
                    id: Ident {
                        id: String::from("GetJoke"),
                    },
                    resource: Ident {
                        id: String::from("JokeService"),
                    },
                    body: vec![Prop {
                        key: String::from("output"),
                        value: Value::Map(vec![Prop {
                            key: String::from("joke"),
                            value: Type::String,
                            decorators: None,
                        }]),
                        decorators: None,
                    }],
                }),
            ],
        };

        let ast = parser.parse(lexer).unwrap();

        assert_eq!(ast, expected_ast);
    }
    #[test]
    pub fn _01_simple_parsing_works() {
        let source = std::fs::read_to_string("test-schemas/01/01.omni").unwrap();
        let lexer = Lexer::new(&source);
        let parser = omni::OmniFileParser::new();

        let expected_ast = File {
            imports: None,

            body: vec![
                Def::Service {
                    id: Ident {
                        id: String::from("JokeService"),
                    },
                    body: vec![Prop {
                        key: String::from("version"),
                        value: Value::String(String::from("1.0.0")),
                        decorators: None,
                    }],
                },
                Def::Read(ResourceOperation {
                    id: Ident {
                        id: String::from("GetJoke"),
                    },
                    resource: Ident {
                        id: String::from("JokeService"),
                    },
                    body: vec![Prop {
                        key: String::from("output"),
                        value: Value::Map(vec![Prop {
                            key: String::from("joke"),
                            value: Type::String,
                            decorators: None,
                        }]),
                        decorators: None,
                    }],
                }),
            ],
        };

        let ast = parser.parse(lexer).unwrap();

        assert_eq!(ast, expected_ast);
    }
}
