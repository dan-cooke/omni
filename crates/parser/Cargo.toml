[package]
name = "parser"
version = "0.1.0"
authors = ["Daniel Cooke <danielcooke1996@gmail.com"]
edition = "2021"

[build-dependencies]
lalrpop = "0.20.0"

[dependencies]
lalrpop-util = { version = "0.20.0", features = ["lexer"] }
logos = "0.13.0"
serde = { version = "1.0.9", features = ["derive"] }

serde_json = { version = "1.0.9" }

[dev-dependencies]
insta = { version = "1.34.0", features = ["json"] }




