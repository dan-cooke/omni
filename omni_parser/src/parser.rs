lalrpop_mod!(omni);
use crate::{ast::File, lexer::Lexer};
use lalrpop_util::lalrpop_mod;
use omni::*;

/// Parse an Omni file
/// This is just a placeholder for the public API for now
pub fn parse_file(file_path: &str) -> File {
    let source = std::fs::read_to_string(file_path)
        .expect(format!("Could not find file {}", file_path).as_str());
    let lexer = Lexer::new(&source);
    let omni_parser = FileParser::new();
    omni_parser.parse(file_path, lexer).unwrap()
}
