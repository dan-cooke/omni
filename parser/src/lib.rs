pub mod ast;
mod errors;
mod helpers;
mod lexer;
mod parser;
mod tokens;

pub use parser::*;
