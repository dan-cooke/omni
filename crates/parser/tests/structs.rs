use insta::assert_json_snapshot;
use parser::*;

#[test]
pub fn structs_can_be_parsed() {
    let ast = parse_file("tests/test-schemas/types/structs.omni");

    assert_json_snapshot!(ast);
}
