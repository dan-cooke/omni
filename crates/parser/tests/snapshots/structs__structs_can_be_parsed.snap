---
source: crates/parser/tests/structs.rs
expression: ast
---
{
  "body": [
    {
      "StructDef": {
        "id": {
          "name": "MyFirstStruct"
        },
        "properties": [
          {
            "id": {
              "name": "foo"
            },
            "value": {
              "Literal": {
                "String": "\"bar\""
              }
            }
          },
          {
            "id": {
              "name": "nested"
            },
            "value": {
              "ObjectExpression": {
                "properties": [
                  {
                    "id": {
                      "name": "fields"
                    },
                    "value": {
                      "Literal": {
                        "String": "\"hello\""
                      }
                    }
                  },
                  {
                    "id": {
                      "name": "nested_again"
                    },
                    "value": {
                      "ObjectExpression": {
                        "properties": [
                          {
                            "id": {
                              "name": "number"
                            },
                            "value": {
                              "Literal": {
                                "Integer": 1234
                              }
                            }
                          },
                          {
                            "id": {
                              "name": "decimal"
                            },
                            "value": {
                              "Literal": {
                                "Float": 12.23
                              }
                            }
                          }
                        ]
                      }
                    }
                  }
                ]
              }
            }
          },
          {
            "id": {
              "name": "identifiers"
            },
            "value": {
              "Identifier": {
                "name": "AnotherStruct"
              }
            }
          }
        ]
      }
    },
    {
      "StructDef": {
        "id": {
          "name": "AnotherStruct"
        },
        "properties": [
          {
            "id": {
              "name": "id"
            },
            "value": {
              "Literal": {
                "String": "\"hello\""
              }
            }
          }
        ]
      }
    }
  ]
}
